src/main/java/com/google/vinet/data/TimeSlot.java

```diff
@@ -0,0 +1,118 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.vinet.data;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+public abstract class TimeSlot implements Comparable<TimeSlot> {
+
+  private final Instant start;
+  private final Instant end;
+  protected final RegisteredUser registeredUser;
+  private final Set<TimeSlot> neighbours = new HashSet<>();
+  private TimeSlot pairedSlot = null;
+  private double distance = 0;
+
+  /**
+   * Create a new time slot. Start and end times must not be null, and start time must come before
+   * end time.
+   *
+   * @param start          The starting instant for the slot.
+   * @param end            The ending instant for the slot.
+   * @param registeredUser The user for which this slot applies.
+   */
+  public TimeSlot(Instant start, Instant end, RegisteredUser registeredUser) {
+    this.start = Objects.requireNonNull(start, "Start time must not be null!");
+    this.end = Objects.requireNonNull(end, "End time must not be null!");
+    if (start.isAfter(end)) throw new IllegalArgumentException();
+    this.registeredUser = registeredUser;
+  }
+
+  public Instant getStart() {
+    return start;
+  }
+
+  public Instant getEnd() {
+    return end;
+  }
+
+  public double getDistance() {
+    return distance;
+  }
+
+  public void setDistance(double distance) {
+    this.distance = distance;
+  }
+
+  public boolean isPaired() {
+    return pairedSlot != null;
+  }
+
+  public TimeSlot getPairedSlot() {
+    return pairedSlot == null ? MatchingAlgorithm.NIL_NODE : pairedSlot;
+  }
+
+  public void setPairedSlot(TimeSlot pairedSlot) {
+    this.pairedSlot = pairedSlot;
+  }
+
+  public Set<TimeSlot> getNeighbours() {
+    return Collections.unmodifiableSet(neighbours);
+  }
+
+  public void addNeighbour(TimeSlot timeSlot) {
+    neighbours.add(timeSlot);
+  }
+
+  /** Compares two time slots based on their start time */
+  @Override
+  public int compareTo(TimeSlot timeSlot) {
+    return start.compareTo(timeSlot.getStart());
+  }
+
+  public static Comparator<TimeSlot> TimeSlotEndComparator = Comparator.comparing(TimeSlot::getEnd);
+
+  /**
+   * Checks whether this timeslot contains another
+   *
+   * @param timeSlot The slot that is to be contained
+   * @return Whether this slot contains the specified slot
+   */
+  public boolean contains(TimeSlot timeSlot) {
+    return compareTo(timeSlot) <= 0 && TimeSlotEndComparator.compare(this, timeSlot) >= 0;
+  }
```

Comment:
I think this piece of code can be simplified.

Instead of breaking the check like this, why not just check if s1 <= s2 and e2 <= e1?

Reply:
I had implemented the compareTo method with the intention of using TimeSlots in sorted TreeSet for the algorithm, but that proved not to be necessary. I thought I would use the already existing APIs rather than do the checks again, but I can change it if it would make the code clearer.

Reply:
Ya, cleaning this up a bit is fine.